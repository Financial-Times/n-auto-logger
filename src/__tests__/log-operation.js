import logger from '../index';
import logOperation from '../log-operation';

jest.mock('@financial-times/n-logger');

describe('logOperation', () => {
	afterEach(() => {
		jest.resetAllMocks();
	});

	describe('can enhance sync operationFunction', () => {
		const monitorFunction = jest.fn();
		const errorInstance = { status: 404, message: 'Not Found' };
		const operationFunction = (req, res) => {
			monitorFunction(req.meta);
			if (req.error) {
				throw errorInstance;
			}
			res.send();
		};
		const enhancedOperation = logOperation(operationFunction);
		const mockRes = {
			send: jest.fn(),
		};
		const errorReq = {
			error: true,
		};

		it('to log operationFunction name as operation name', () => {
			enhancedOperation({}, mockRes);
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('to invoke operationFunction with operationFunction name in req.meta.operation', () => {
			enhancedOperation({}, mockRes);
			expect(monitorFunction.mock.calls).toMatchSnapshot();
			expect(mockRes.send.mock.calls).toMatchSnapshot();
		});

		it('to log operaitonFunction on success', () => {
			enhancedOperation({}, mockRes);
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('inherit meta passed to req.meta from previous middleware and log them', () => {
			const reqWithMeta = { meta: { userId: 'mock-id' } };
			enhancedOperation(reqWithMeta, mockRes);
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('to log operationFunction on failure and throw the caught error', async () => {
			try {
				await enhancedOperation(errorReq, mockRes);
			} catch (e) {
				expect(e).toBe(errorInstance);
				expect(logger.error.mock.calls).toMatchSnapshot();
				expect(logger.warn.mock.calls).toMatchSnapshot();
			}
		});
	});

	describe('can enhance async operationFunction', () => {
		const monitorFunction = jest.fn();
		const errorInstance = { status: 404, message: 'Not Found' };
		const operationFunction = async (req, res) => {
			await monitorFunction(req.meta);
			if (req.error) {
				throw errorInstance;
			}
			res.send();
		};
		const enhancedOperation = logOperation(operationFunction);
		const mockRes = {
			send: jest.fn(),
		};
		const errorReq = {
			error: true,
		};

		it('to log operationFunction name as operation name', async () => {
			await enhancedOperation({}, mockRes);
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('to invoke operationFunction with operationFunction name in req.meta.operation', async () => {
			await enhancedOperation({}, mockRes);
			expect(monitorFunction.mock.calls).toMatchSnapshot();
			expect(mockRes.send).toHaveBeenCalled();
		});

		it('to log operaitonFunction on success', async () => {
			await enhancedOperation({}, mockRes);
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('inherit meta passed to req.meta from previous middleware and log them', async () => {
			const reqWithMeta = { meta: { userId: 'mock-id' } };
			await enhancedOperation(reqWithMeta, mockRes);
			expect(logger.info.mock.calls).toMatchSnapshot();
		});

		it('to log operationFunction on failure and throw the caught error', async () => {
			try {
				await enhancedOperation(errorReq, mockRes);
			} catch (e) {
				expect(e).toBe(errorInstance);
				expect(logger.error.mock.calls).toMatchSnapshot();
				expect(logger.warn.mock.calls).toMatchSnapshot();
			}
		});
	});
});
