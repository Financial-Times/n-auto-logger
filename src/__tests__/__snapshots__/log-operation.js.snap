// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logOperation can enhance async operationFunction inherit meta passed to req.meta from previous middleware and log them 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
      "userId": "mock-id",
    },
  ],
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
      "userId": "mock-id",
    },
  ],
]
`;

exports[`logOperation can enhance async operationFunction to invoke operationFunction with operationFunction name in req.meta.operation 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
]
`;

exports[`logOperation can enhance async operationFunction to log operaitonFunction on success 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
    },
  ],
]
`;

exports[`logOperation can enhance async operationFunction to log operationFunction name as operation name 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
    },
  ],
]
`;

exports[`logOperation can enhance async operationFunction to log operationFunction on failure and throw the caught error 1`] = `Array []`;

exports[`logOperation can enhance async operationFunction to log operationFunction on failure and throw the caught error 2`] = `
Array [
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;

exports[`logOperation can enhance sync operationFunction inherit meta passed to req.meta from previous middleware and log them 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
      "userId": "mock-id",
    },
  ],
]
`;

exports[`logOperation can enhance sync operationFunction to invoke operationFunction with operationFunction name in req.meta.operation 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
]
`;

exports[`logOperation can enhance sync operationFunction to log operaitonFunction on success 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
]
`;

exports[`logOperation can enhance sync operationFunction to log operationFunction name as operation name 1`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
    },
  ],
]
`;

exports[`logOperation can enhance sync operationFunction to log operationFunction on failure and throw the caught error 1`] = `Array []`;

exports[`logOperation can enhance sync operationFunction to log operationFunction on failure and throw the caught error 2`] = `
Array [
  Array [
    Object {
      "category": "CUSTOM_ERROR",
      "message": "Not Found",
      "operation": "operationFunction",
      "result": "failure",
      "status": 404,
    },
  ],
]
`;
